### DISCUSSION TITLE
[DAY 7]피어세션

### DISCUSSION THREAD

**Discussion by karim-moon** [bcaitech1/p4-fr-soccer#47] [Peersession]
- 정환
    - focal loss, cosine focal loss 모듈 추가(loss.py)
    - ASTER에 focal loss(alpha=1, gamma=2, ignore_index=pad) 적용 후 학습중
    - [https://github.com/Media-Smart/vedastr](https://github.com/Media-Smart/vedastr) 구현중
    - beam search 구현중..
- 문종
    - Positional Encoder SATRN에 논문처럼 재현(똑같이 재현X)
    - ASTER Efficient net 학습완료 → 제출 스코어 (0.6403)
    - 50epoch부터 resume하여 재학습중
- 수호
    - ASTER 4:1 비율로 학습 시켜보기(현재 괜찮은 성능을 보이고 있음)
    - SATRN에 Locality-aware feedforward layer 논문 구현
    - ignore_index에 대한 차이 확인하기(nn.crossentropy())
        - 문제가 많이 발생했음. <PAD> 토큰 실종.

![EOS](https://user-images.githubusercontent.com/52662993/120442691-dd5e1600-c3c0-11eb-876d-caea8d658fde.png)
    - feed forward에 대한 비교(separable이 좋은 것 같다)

  ![Graph](https://user-images.githubusercontent.com/52662993/120442590-bef81a80-c3c0-11eb-8857-853a752f2aad.png)

- 동빈
- 근재
    - EfficientNet 적용한 SATRN 학습중
    - SATRN에 focal loss 적용하기
    - ignore_index 적용 후 학습

# PAD token을 loss의 ignore index로 잡을 때 차이

nn.CrossEntropyLoss(ignore_index=train_dataset.token_to_id[PAD]) 한다면..

GT     =  1     a        <PAD> 

PRED =  1 <PAD>  <PAD>  인 경우, 

GT     = [ 0 0 ... 1 ... 0 0 ] [ 0 0 0 0 0 0 ... 1 ... 0 ] [ 0 0 0 1 ... 0 ]
PRED = [ 0 0 ... 1 ... 0 0 ] [ 0 0 0 1 0 0 ... 0 ... 0 ] [ 0 0 0 1 ... 0 ]

으로 <PAD>가 불일치 하는 값에 대해 기존 loss보다 해당 수치만 반 줄어드는 듯? 

# 35~ 45 epoch정도부터 Focal loss를 사용하면 어떨까

Wandb 그래프상 35 ~ 45 epoch정도 학습이 진행되었을 때 스코어가 더디게 올라가는 경향이 있어 이때부터 Loss를 Crossentropy → Focal Loss로 바꿔준다면 성능이 오를것으로 기대

# SATRN 베이스라인 논문 재현

베이스라인에 2D positional Encoder, Feed Foward부분 재현이 안돼있는상태

- [x]  2D PE 재현
- [x]  Feed Foward 재현

<img width="718" alt="conv" src="https://user-images.githubusercontent.com/52662993/120442803-f961b780-c3c0-11eb-90ad-487a066ea915.png">

- (a), (b), (c) option중 깃에 참고할 수 있는 구조는 (b), 추후 (c) 구현하는 방향으로
